src/metatrain/experimental/nanopet/model.py

```diff
@@ -0,0 +1,474 @@
+from ...utils.data import DatasetInfo, TargetInfo
+from ...utils.dtype import dtype_to_str
+from .modules.encoder import Encoder
+from .modules.nef import (
+    edge_array_to_nef,
+    get_corresponding_edges,
+    get_nef_indices,
+    nef_array_to_edges,
+)
+from .modules.radial_mask import get_radial_mask
+from .modules.structures import concatenate_structures
+from .modules.transformer import Transformer
+
+
+class NanoPET(torch.nn.Module):
+
+    __supported_devices__ = ["cuda", "cpu"]
+    __supported_dtypes__ = [torch.float64, torch.float32]
+
+    component_labels: Dict[str, List[Labels]]
+
+    def __init__(self, model_hypers: Dict, dataset_info: DatasetInfo) -> None:
+        super().__init__()
+
+        for target in dataset_info.targets.values():
+            if target.is_spherical:
+                raise ValueError(
+                    "The NanoPET model does not support spherical tensor targets. "
+                    "Only scalar and Cartesian tensor targets are supported."
+                )
+
+        self.hypers = model_hypers
+        self.dataset_info = dataset_info
+        self.new_outputs = list(dataset_info.targets.keys())
+        self.atomic_types = dataset_info.atomic_types
+
+        self.cutoff = self.hypers["cutoff"]
+        self.cutoff_width = self.hypers["cutoff_width"]
+
+        self.encoder = Encoder(len(self.atomic_types), self.hypers["d_pet"])
+
+        self.transformer = Transformer(
+            self.hypers["d_pet"],
+            4 * self.hypers["d_pet"],
+            self.hypers["num_heads"],
+            self.hypers["num_attention_layers"],
+            0.0,  # MLP dropout rate
+            0.0,  # attention dropout rate
+        )
+        # empirically, the model seems to perform better without dropout
+
+        self.num_mp_layers = self.hypers["num_gnn_layers"] - 1
+        gnn_contractions = []
+        gnn_transformers = []
+        for _ in range(self.num_mp_layers):
+            gnn_contractions.append(
+                torch.nn.Linear(
+                    2 * self.hypers["d_pet"], self.hypers["d_pet"], bias=False
+                )
+            )
+            gnn_transformers.append(
+                Transformer(
+                    self.hypers["d_pet"],
+                    4 * self.hypers["d_pet"],
+                    self.hypers["num_heads"],
+                    self.hypers["num_attention_layers"],
+                    0.0,  # MLP dropout rate
+                    0.0,  # attention dropout rate
+                )
+            )
+        self.gnn_contractions = torch.nn.ModuleList(gnn_contractions)
+        self.gnn_transformers = torch.nn.ModuleList(gnn_transformers)
+
+        self.last_layer_feature_size = self.hypers["d_pet"]
+
+        # register the outputs
+        # the model is always capable of outputting the last layer features
+        self.outputs = {
+            "mtt::aux::last_layer_features": ModelOutput(unit="unitless", per_atom=True)
```

Comment:
Should this be updated? Or is this a separate PR?

```suggestion
            "features": ModelOutput(unit="unitless", per_atom=True)
```

Reply:
👍